### PR TITLE
fix(cluster-link): correctly replicate overlapping route intersections

### PR DESCRIPTION
**Pull request by keynslug**
Fixes [EMQX-13229](https://emqx.atlassian.net/browse/EMQX-13229)

Release version: e5.8.2

## Summary

Cluster Linking involves route replication and allows to configure a _set of topic filters_ linked cluster is interested in. This set of topic filters essentially represents a subset of topic space, and it's currently infeasible to always have exact and minimal representation of such set: 1) a topic filter in this set can be a "subset" of another topic filter, or 2) a pair of topic filters can be overlapping.

For the first case this PR introduces config validation that rejects such redundant topic filters.

The second case is worse though: if route intersects with 2 topic filters that are overlapping, we may have to replicate this route 2 times, in cases where these intersections are themselves overlapping. Before these changes, the route replication assumed that it's enough to replicate just a single non-empty intersection. Usually the first one would be replicated: e.g. with `t/+` subscription and `[t/1, t/#]` link filters, only `t/1` would be the only replicated route.

This PR also refactors `emqx_retainer` slightly, because the responsibility of filtering out special topics became split between the backend and the dispatcher after `emqx_topic:match/2` refinements. This responsibility is now exclusively in the backend, where it's also cheaper to do.

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [ ] ~~Added property-based tests for code which performs user input validation~~
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

[EMQX-13229]: https://emqx.atlassian.net/browse/EMQX-13229?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/emqx/src/emqx_topic.erl [hunk 1]
```diff
@@ -150,6 +168,29 @@ intersect_join(W, Words) -> [W | Words].
 wildcard_intersection(W, W) -> W;
 wildcard_intersection(_, _) -> '+'.
 
+%% @doc Finds out if topic / topic filter T1 is a subset of topic / topic filter T2.
+-spec subset(topic() | words(), topic() | words()) -> boolean().
```
Comment (zmstone) [inline apps/emqx/src/emqx_topic.erl:172]: is_subset ?

**Review by zmstone — APPROVED** [review]

**Comment by savonarola** [comment]
Just curious: what is the summary of the original logic problem? I.e. from the first view it's unclear why overlapping routes prevented messages from being published (as the related issue states).

**Comment by keynslug** [comment]
> Just curious: what is the summary of the original logic problem?

Before these changes, the route replication assumed that it's enough to replicate just a single non-empty intersection. Usually the first one would be replicated: e.g. with `t/+` subscription and `[t/1, t/#]` link filters, only `t/1` would be the only replicated route.

Added this into the PR description.

File: apps/emqx_retainer/test/emqx_retainer_SUITE.erl
Comment (savonarola) [inline apps/emqx_retainer/test/emqx_retainer_SUITE.erl:1]: NB: This test is also used by fdb retainer

File: apps/emqx_retainer/src/emqx_retainer_mnesia.erl [hunk 2]
```diff
@@ -381,12 +381,24 @@ search_stream(Tokens, Now) ->
...
     Ms = make_message_match_spec(Tokens, Now),
-    emqx_utils_stream:ets(
+    MsgStream = emqx_utils_stream:ets(
         fun
             (undefined) -> ets:select(?TAB_MESSAGE, Ms, 1);
             (Cont) -> ets:select(Cont)
         end
-    );
+    ),
+    case Tokens of
```
Comment (savonarola) [inline apps/emqx_retainer/src/emqx_retainer_mnesia.erl:390]: Maybe a corresponding update is needed for the fdb retainer

**Review by savonarola — APPROVED** [review]

File: apps/emqx_retainer/src/emqx_retainer_mnesia.erl (hunk 2 above)
Comment (keynslug) [reply apps/emqx_retainer/src/emqx_retainer_mnesia.erl:390]: Will take care of that once the next sync PR is merged.

**Review by zmstone — APPROVED** [review]

**Comment by emqxqa** [comment]
[TestExecution](https://emqx.atlassian.net/browse/QA-4775)

**Pull request merged**